More SAMx7x ADC features

chris-durand opened the pull request:
- [x] Oversampling
- [x] Multi-channel conversion sequences
- [x] Hardware triggering
- [x] Examples
  - [x] Multi-channel sequence
  - [x] DMA with hardware timer trigger
- [x] Test examples in hardware

chris-durand committed:
[sam] Add SAMx7x ADC hardware trigger and oversampling

chris-durand committed:
[sam] Add SAMx7x ADC multi-channel conversion sequences

chris-durand committed:
[example] Add SAMx7x ADC multi-channel sequence example

chris-durand committed:
[example] Add SAMx7x ADC DMA example with hardware timer trigger

salkinium commented on examples/samv71_xplained_ultra/adc/dma/main.cpp:
```diff
@@ -0,0 +1,134 @@
+ */
+
+using namespace modm::platform;
+using namespace modm::literals;
```
Isn't this already in the board.hpp?

chris-durand replied:
In board.hpp we have:
```cpp
namespace Board
{
/// @ingroup modm_board_samv71_xplained_ultra
/// @{
using namespace modm::literals;
using namespace modm::platform;
```

This is much more sensible and friendly to users than having it in global scope by default in the board header.

salkinium replied:
Hm, interesting, it's inconsistent with [previous boards](https://github.com/modm-io/modm/blob/develop/src/modm/board/nucleo_f103rb/board.hpp#L22). I think I remember a discussion about exactly this a long time ago and therefore the `modm::literals` is NOT top-level. We should probably make it consistent at some time, even if we have to refactor the examples.

chris-durand replied:
Yeah, we should probably do that. I'd also argue that as a library we should not put things in the global namespace that could lead to issues for users.

The SAM headers are a good example why we shouldn't do that. They define structs and macros with peripheral names in the global namespace. In the DMA driver I've tried to name a class like the peripheral in the data sheet, `Xdmac`. I had to give up because of ambiguities and renamed it `Dma`.

Thankfully Atmel mostly uses such awkward naming that identifiers rarely collide :smile:, like `Afec` for ADC, `Acc` for comparator and `Twihs` for I²C ...

salkinium commented on examples/samv71_xplained_ultra/adc/multi-channel/main.cpp:
```diff
@@ -0,0 +1,51 @@
+#include <modm/board.hpp>
+#include <modm/io/iostream.hpp>
+
+using namespace modm::platform;
```
also part of board.hpp

salkinium reviewed (approved):
Yes!

salkinium commented on src/modm/platform/adc/samg/adc.hpp.in:
```diff
@@ -198,13 +328,46 @@ public:
+	enableHardwareTrigger(Trigger trigger)
+	{
+		Regs()->{{reg}}_MR &= ~({{reg}}_MR_TRGEN | AFEC_MR_TRGSEL_Msk);
+		Regs()->{{reg}}_MR |= {{reg}}_MR_TRGEN | static_cast<uint32_t>(trigger);
```
Will this not generate two loads/stores because of volatile?

chris-durand replied:
That's intended. Turn it off first, then change it. I am not 100% percent sure, though, if the hardware really requires this.

rleh reviewed (approved):
Nice 👍🏽

salkinium commented:
@chris-durand Please rebase and merge when you have time.

chris-durand commented:
I'll merge it tomorrow.

chris-durand committed:
[sam] Fix SAMx7x ADC readChannel() if previous data has not been read

Clear data ready flag and do not return previous result if it has not
been read from the LCDR register. When using conversion sequences data
is normally read from the CDR register which does not clear the data
ready flag.

chris-durand commented:
I have pushed another commit to fix a driver bug.

rleh reviewed (approved):
Go!
